src/Middleware/UnauthenticatedRedirectMiddleware.php

```diff
@@ -0,0 +1,65 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         1.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Authentication\Middleware;
+
+use Authentication\AuthenticationService;
+use Authentication\Authenticator\UnauthorizedException;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Zend\Diactoros\Stream;
+
+/**
+ * Catches UnauthenticatedException and converts
+ * them into redirects. The redirect target is provided as a constructor
+ * argument.
+ */
+class UnauthenticatedRedirectMiddleware
+{
+    /**
+     * The target URL to redirect to.
+     *
+     * @var string
+     */
+    protected $target;
+
+    /**
+     * Constructor
+     *
+     * @param string $target The target URL that redirects should be sent to.
+     */
+    public function __construct($target)
+    {
+        $this->target = $target;
+    }
+
+    /**
+     * Callable implementation for the middleware stack.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request.
+     * @param \Psr\Http\Message\ResponseInterface $response The response.
+     * @param callable $next The next middleware to call.
+     * @return \Psr\Http\Message\ResponseInterface A response.
+     */
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, $next)
+    {
+        try {
+            return $next($request, $response);
+        } catch (UnauthorizedException $e) {
```

Comment:
Won't this redirect HTTP authentication as well?

Reply:
It could, but if you stack the middleware correctly it won't.